X2WOTCCommunityHighlander/Src/XComGame/Classes/X2AbilityTrigger_EndOfMove.uc

```
@@ -27,25 +33,99 @@ static function bool AbilityHasEndOfMoveTrigger(X2AbilityTemplate Template)
+///
+/// ```
+/// local X2AbilityTrigger_EndOfMove EndOfMoveTrigger;
+/// EndOfMoveTrigger = new class'X2AbilityTrigger_EndOfMove';
+/// EndOfMoveTrigger.RClickMeleePriority = <<CUSTOM INT VALUE>>;
+/// Template.AbilityTriggers.AddItem(EndOfMoveTrigger);
+/// ```
 static function XComGameState_Ability GetAvailableEndOfMoveAbilityForUnit(XComGameState_Unit UnitState)
 {
 	local XComGameStateHistory History;
 	local GameRulesCache_Unit UnitCache;
-	local XComGameState_Ability AbilityState;
-	local int ActionIndex;
+	local AvailableAction Action;
+	local XComGameState_Ability AbilityState, PriorityAbility;
+	local X2AbilityTrigger_EndOfMove EndOfMoveTrigger;
+	local int CurrentPriority, AbilityPriority;
+
+	PriorityAbility = none;
+	CurrentPriority = 2147483647; // Max INT
 
 	if(`TACTICALRULES.GetGameRulesCache_Unit(UnitState.GetReference(), UnitCache))
 	{
 		History = `XCOMHISTORY;
-		for(ActionIndex = 0; ActionIndex < UnitCache.AvailableActions.Length; ActionIndex++)
+		foreach UnitCache.AvailableActions(Action)
 		{
-			AbilityState = XComGameState_Ability(History.GetGameStateForObjectID(UnitCache.AvailableActions[ActionIndex].AbilityObjectRef.ObjectID));
-			if(AbilityState != none && AbilityHasEndOfMoveTrigger(AbilityState.GetMyTemplate()))
+			// Is the action even available?
+			if(Action.AvailableCode != 'AA_Success')
+			{
+				continue;
+			}
+
+			AbilityState = XComGameState_Ability(History.GetGameStateForObjectID(Action.AbilityObjectRef.ObjectID));
+			// Unknown ability, so move on
+			if(AbilityState == none)
+			{
+				continue;
+			}
+
+			EndOfMoveTrigger = GetEndOfMoveTrigger(AbilityState.GetMyTemplate());
+			// Not right-click enabled, so move on
+			if(EndOfMoveTrigger == none)
+			{
+				continue;
+			}
+
+			// Calculate priority
+			// If default value then use SHOT_HUD_PRIORITY, otherwise grab its priority
+			AbilityPriority = EndOfMoveTrigger.RClickMeleePriority >= 0 ? EndOfMoveTrigger.RClickMeleePriority : const.SHOT_HUD_PRIORITY;
```

Several questions:

1) Why `SHOT_HUD_PRIORITY` was chosen as the name for default right-click-melee priority? Seems confusing/misleading to me
2) Why are negative custom explicit priorities not supported? e.g. in Nonlethal Combat case, wouldn't it make sense to set Subdue to something negative (very low) so that any other melee ability trumps it?

---

Negative values are not supported because it is mirroring the ShotHUDPriority scale, meaning that 0 will be the highest priority due to wanting the "left-most" skill in the Shot HUD.  I have tweaked the comments and naming conventions to make this clearer.